packages/broker/src/plugins/operator/MaintainOperatorValueHelper.ts

```diff
@@ -1,33 +1,127 @@
+        // filter out my own operator
+        const operatorAddresses = operators.filter((id) => id !== this.config.operatorContractAddress)
+        logger.info(`Found ${operatorAddresses.length} operators`, { operatorAddresses })
+        const randomIndex = Math.floor(Math.random() * operatorAddresses.length)
+        return operatorAddresses[randomIndex]
+    }
+
+    /**
+     * The "hard limit" for paying out rewards to `withdrawEarningsFromSponsorships` caller.
+     * Operator is expected to call `withdrawEarningsFromSponsorships` before
+     *   `unwithdrawn earnings / (total staked + free funds)` exceeds this limit.
+     * @returns a "wei" fraction: 1e18 or "1 ether" means limit is at unwithdrawn earnings == total staked + free funds
+     */
     async getPenaltyLimitFraction(): Promise<bigint> {
         const streamrConfigAddress = await this.operator.streamrConfig()
         const streamrConfig = new Contract(streamrConfigAddress, streamrConfigABI, this.config.provider) as unknown as StreamrConfig
         return (await streamrConfig.poolValueDriftLimitFraction()).toBigInt()
     }
 
-    // ethers5 handles BigIng as BigNumber, but once it's upgrated to ethers6, it will be changed to BigInt (see ETH-536)
-    async getApproximatePoolValuesPerSponsorship(): Promise<{
-        sponsorshipAddresses: string[]
-        approxValues: BigNumber[]
-        realValues: BigNumber[]
-    }> {
-        return await this.operator.getApproximatePoolValuesPerSponsorship()
+    /**
+     * Find the sum of unwithdrawn earnings in Sponsorships (that the Operator must withdraw before the sum reaches a limit),
+     * SUBJECT TO the constraints, set in the OperatorServiceConfig:
+     *  - only take at most maxSponsorshipsCount addresses (those with most earnings), or all if undefined
+     *  - only take sponsorships that have more than minSponsorshipEarnings, or all if undefined
+     * @param operatorContractAddress
+     */
+    async getUnwithdrawnEarningsOf(operatorContractAddress: EthereumAddress): Promise<UnwithdrawnEarningsData> {
+        const operator = new Contract(operatorContractAddress, operatorABI, this.config.signer) as unknown as Operator
+        const minSponsorshipEarningsWei = BigNumber.from(this.config.minSponsorshipEarnings ?? 0)
+        const { sponsorshipAddresses: allSponsorshipAddresses, earnings } = await operator.getEarningsFromSponsorships()
+
+        const sponsorships = allSponsorshipAddresses
+            .map((address, i) => ({ address, earnings: earnings[i] }))
+            .filter((sponsorship) => sponsorship.earnings.gte(minSponsorshipEarningsWei))
+            .sort((a, b) => Number(b.earnings.sub(a.earnings).toBigInt())) // TODO: after Node 20, use .toSorted() instead
+            .slice(0, this.config.maxSponsorshipsCount) // take all if maxSponsorshipsCount is undefined
+        const sponsorshipAddresses = sponsorships.map((sponsorship) => sponsorship.address as EthereumAddress)
```

Comment:
If you use the `EthereumAddress` type it is assumed that it is (a) a valid address and (b) in lower-case. Here the first one most likely always holds since it is read from a smart contract but the second one not.

So we should avoid casting here and instead use the helper function `toEthereumAddress`:

```suggestion
        const sponsorshipAddresses = sponsorships.map((sponsorship) => toEthereumAddress(sponsorship.address))
```

Reply:
Changed